### ISSUE TITLE
calico-typha memory usage keeps growing

### ISSUE THREAD

**Issue by svend** [projectcalico/typha#303]
## Expected Behavior

The memory usage for calico-typha Kubernetes pods should stabilize.

## Current Behavior

The memory usage for the calico-typha pods starts at under 40 MiB. The memory memory usage grows at about 6 MB per day until it hits the memory limit, which we have set at 100 MiB.

## Steps to Reproduce (for bugs)

We have deployed calico calico using https://docs.projectcalico.org/v3.8/manifests/calico-typha.yaml. We see the memory growth across multiple Kubernetes clusters.

Here is a sample graph of the memory usage over two weeks.

![image](https://user-images.githubusercontent.com/306190/64277016-2a683c00-cefe-11e9-8902-8d08bab36af5.png)

The calico-typha deployment definition is:

``` yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "3"
  creationTimestamp: "2019-02-13T19:37:14Z"
  generation: 4
  labels:
    k8s-app: calico-typha
  name: calico-typha
  namespace: kube-system
  resourceVersion: "55618639"
  selfLink: /apis/extensions/v1beta1/namespaces/kube-system/deployments/calico-typha
  uid: c3718932-2fc6-11e9-8172-02962cb1810e
spec:
  progressDeadlineSeconds: 600
  replicas: 3
  revisionHistoryLimit: 2
  selector:
    matchLabels:
      k8s-app: calico-typha
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      annotations:
        cluster-autoscaler.kubernetes.io/safe-to-evict: "true"
        logging_service: calico-typha
        prometheus.io/port: "9093"
        prometheus.io/scrape: "true"
        scheduler.alpha.kubernetes.io/critical-pod: ""
      creationTimestamp: null
      labels:
        k8s-app: calico-typha
    spec:
      containers:
      - env:
        - name: TYPHA_LOGSEVERITYSCREEN
          value: info
        - name: TYPHA_LOGFILEPATH
          value: none
        - name: TYPHA_LOGSEVERITYSYS
          value: none
        - name: TYPHA_CONNECTIONREBALANCINGMODE
          value: kubernetes
        - name: TYPHA_DATASTORETYPE
          value: kubernetes
        - name: TYPHA_HEALTHENABLED
          value: "true"
        - name: TYPHA_PROMETHEUSMETRICSENABLED
          value: "true"
        - name: TYPHA_PROMETHEUSMETRICSPORT
          value: "9093"
        image: calico/typha:v3.8.0
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            host: localhost
            path: /liveness
            port: 9098
            scheme: HTTP
          initialDelaySeconds: 30
          periodSeconds: 30
          successThreshold: 1
          timeoutSeconds: 1
        name: calico-typha
        ports:
        - containerPort: 5473
          hostPort: 5473
          name: calico-typha
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            host: localhost
            path: /readiness
            port: 9098
            scheme: HTTP
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        resources:
          limits:
            cpu: 100m
            memory: 100Mi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      hostNetwork: true
      nodeSelector:
        beta.kubernetes.io/os: linux
        wp/controller-node: "true"
      priorityClassName: system-cluster-critical
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: calico-node
      serviceAccountName: calico-node
      terminationGracePeriodSeconds: 30
      tolerations:
      - key: CriticalAddonsOnly
        operator: Exists
      - effect: NoSchedule
        key: wp-dedicated
        value: control-plane
```

## Your Environment

* Typha version (if using Typha binary directly `calico-typha --version`):
* And/Or, `calico/typha` container version (if running Typha in a container): v3.8.0
* Which orchestrator are you using (e.g. Kubernetes, OpenStack, Docker, Mesos): Kubernetes
* Orchestrator version (if used): v1.14.5
* Etcd version (if using etcd datastore driver):
* Operating System and version:
* Link to your project (optional):

**Comment by spikecurtis** [comment]
I was wondering if you could share some more information about your cluster to help us track down where this memory might be going.

Is your cluster changing size on any kind of regular basis? E.g. autoscaling, worker nodes joining and leaving, etc.

Are you modifying network policy on a regular basis? (Kubernetes NetworkPolicy, Calico NetworkPolicy, GlobalNetworkPolicy)

Are you using/changing Network Sets?

Are you using/changing HostEndpoints?

**Comment by svend** [comment]
Hi @spikecurtis. Thanks for your help!

> Is your cluster changing size on any kind of regular basis? E.g. autoscaling, worker nodes joining and leaving, etc.

Our clusters are autoscaled using the cluster-autoscaler. Our clusters (which are about 30 nodes each) have about 1-3 nodes added and removed by the autoscaler per day.

> Are you modifying network policy on a regular basis? (Kubernetes NetworkPolicy, Calico NetworkPolicy, GlobalNetworkPolicy)

I checked our clusters, and didn't find any resources of those types defined.

> Are you using/changing Network Sets?

We don't have any NetworkSet resources.

> Are you using/changing HostEndpoints?

We don't have any HostEndpoints resources.

**Comment by spikecurtis** [comment]
@svend I'm not able to replicate this in our environment yet.

Would you be willing to run a modified version of Typha in your environment that periodically dumps a memory pprof that I could examine to look for leaks?

**Comment by spikecurtis** [comment]
@svend I'm unable to reproduce this behavior.  Can you share any typha log files, or any further details that might shed some light on what's different about your environment?

**Comment by svend** [comment]
@spikecurtis I can run the modified typa container.

I believe we also captured some pprof dumps from our deployment. I'll see if I can dig those up.

**Comment by spikecurtis** [comment]
pprof dumps are exactly what I need! If you can find them please send them my way.  In particular, I'm looking for the heap profile.  If you can't find them, or didn't take a heap profile...

It turns out **fasaxc** cleverly already added code to dump the heap profile, so no modified binary is needed. It is a little involved because Typha is built with a scratch container, but what you'll need to do is add the following environment variable to your Typha manifest

```
          - name: TYPHA_DEBUGMEMORYPROFILEPATH
            value: "/code/heap_<timestamp>"
```

(`/code/` is a directory that already exists in the Typha image).

Adding that environment variable will restart Typha pods, so let it stew for a while and memory usage increase.

Next, you'll need to trigger the dump by sending the `USR1` signal to the process. You'll have to do this from the node, since there isn't a shell to `kubectl exec` into in the Typha container.  Find out which node Typha is on, then ssh to it.  You can learn the `calico-typha` PID by running

```
$ ps -eaf | grep typha

root     21401 21384  0 17:26 ?        00:00:00 /sbin/tini -- calico-typha
root     21430 21401  0 17:26 ?        00:00:00 calico-typha
ubuntu   22112 21941  0 17:27 pts/0    00:00:00 grep --color=auto typha
```

Then send the signal to the PID (in this case 21430)

```
sudo kill -s SIGUSR1 21430
```

Next you'll need to copy the dump out of the container.  The `<timestamp>` in the dump path gets converted to an actual timestamp, so we need to be a little clever to learn the full path. It's in Typha's logs.

E.g.
```
kubectl logs -n kube-system calico-typha-795ddb997d-j87kk
```

You're looking for lines like these
```
2020-02-05 17:27:27.037 [INFO][7] daemon.go 427: Received SIGUSR1, emitting heap profile
2020-02-05 17:27:27.037 [INFO][7] daemon.go 471: Asked to create a memory profile. file="/code/heap_<timestamp>"
2020-02-05 17:27:27.037 [INFO][7] daemon.go 493: Writing memory profile... file="/code/heap_2020-02-05-17:27:27"
2020-02-05 17:27:27.044 [INFO][7] daemon.go 500: Finished writing memory profile file="/code/heap_2020-02-05-17:27:27"
```

Again, the Typha container is built from scratch, so `kubectl cp` won't work. You'll need to ssh to the node, and do a `docker ps` to learn the container name. It'll be something like `k8s_calico-typha_calico-typha-795ddb997d-j87kk_kube-system_6809b404-6c0e-4bd4-889f-d8216c13c13b_0` and the command will be `/sbin/tini -- calico-typha`.

Copy the heap profile out like this

```
docker cp k8s_calico-typha_calico-typha-795ddb997d-j87kk_kube-system_6809b404-6c0e-4bd4-889f-d8216c13c13b_0:/code/heap_2020-02-05-17:27:27 .
```

If you're worried about attaching pprof profiles to a public issue, feel free to email me [spike at tigera dot io].

**Comment by svend** [comment]
I have sent you the pprofs we captured in September. Let me know if these are useful, of if I should capture new ones.

**Comment by spikecurtis** [comment]
I found the leak.  It's in golang.org/x/net/http2 of all places, buried pretty deep.  https://github.com/golang/go/issues/21543

The good news is that it is fixed upstream, and that newer patch releases of Calico have the fix. I've verified that the version of golang.org/x/net/http2 that we use to build Typha v3.8.6 has been patched, so you should be able to upgrade to the latest patch release and see this issue resolved. (It's a good idea to keep up with the latest patch releases anyway, for security and high priority bug fixes.)

It's also fixed in Calico v3.12.0 :-)
